share/sql/migrate.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	for res.Next() {
+		var s OldShareState
+		res.Scan(&s.id, &s.recipient, &s.state)
+		if err == nil {
+			ch <- &s
+		} else {
+			fmt.Printf("Error occured for share status%d: %s\n", s.id, err.Error())
+		}
+	}
+}
+
+func workerShare(ctx context.Context, migrator Migrator, ch chan *OldShareEntry) {
+	for share := range ch {
+		handleSingleShare(ctx, migrator, share)
+	}
+}
+
+func workerState(ctx context.Context, migrator Migrator, ch chan *OldShareState) {
+	for state := range ch {
+		handleSingleState(ctx, migrator, state)
+	}
+}
+
+func handleSingleShare(ctx context.Context, migrator Migrator, s *OldShareEntry) {
+	share, err := oldShareToNewShare(ctx, migrator, s)
+	if err != nil {
+		return
+	}
+	// TODO error handling
+	if share.IsShare {
+		migrator.NewDb.Create(&share.Share)
+	} else {
+		migrator.NewDb.Create(&share.Link)
+	}
+}
+
+func handleSingleState(ctx context.Context, migrator Migrator, s *OldShareState) {
+	// case collaboration.ShareState_SHARE_STATE_REJECTED:
+	// 	state = -1
+	// case collaboration.ShareState_SHARE_STATE_ACCEPTED:
+	// 	state = 1
+
+	newShareState := &model.ShareState{
+		ShareID: uint(s.id),
+		Model: gorm.Model{
+			ID: uint(s.id),
+		},
+		User:   s.recipient,
+		Hidden: s.state == -1, // Hidden if REJECTED
+		Synced: true,          // for now, we always sync? or not? TODO
+	}
+	migrator.NewDb.Create(&newShareState)
+}
+
+func oldShareToNewShare(ctx context.Context, migrator Migrator, s *OldShareEntry) (*ShareOrLink, error) {
+	expirationDate, expirationError := time.Parse("2006-01-02 15:04:05", s.Expiration)
+
+	protoShare := model.ProtoShare{
+		Model: gorm.Model{
+			ID:        uint(s.ID),
+			CreatedAt: time.Unix(int64(s.STime), 0),
+			UpdatedAt: time.Unix(int64(s.STime), 0),
+		},
+		UIDOwner:     s.UIDOwner,
+		UIDInitiator: s.UIDInitiator,
+		Description:  s.Description,
+		Permissions:  uint8(s.Permissions),
+		Orphan:       s.Orphan, // will be re-checked later
+		Expiration: datatypes.Null[time.Time]{
+			V:     expirationDate,
+			Valid: expirationError == nil,
+		},
+		ItemType:    model.ItemType(s.ItemType),
+		InitialPath: "", // set later
+		Inode:       s.ItemSource,
+		Instance:    s.Prefix,
+	}
+
+	// Getting InitialPath
+	if !protoShare.Orphan {
+		path, err := migrator.ShareMgr.getPath(ctx, &providerv1beta1.ResourceId{
+			StorageId: protoShare.Instance,
+			OpaqueId:  protoShare.Inode,
+		})
+		if err == nil {
+			protoShare.InitialPath = path
+		} else if errors.Is(err, errtypes.NotFound(protoShare.Inode)) {
+			protoShare.Orphan = true
+		} else {
+			// We do not set, because of a general error
+			fmt.Printf("An error occured for share %d while statting (%s, %s): %s\n", s.ID, protoShare.Instance, protoShare.Inode, err.Error())
+		}
+	}
+
+	// ShareTypeUser = 0
+	// ShareTypeGroup = 1
+	// ShareTypePublicLink = 3
+	// ShareTypeFederatedCloudShare = 6
+	// ShareTypeSpaceMembership = 7
+	if s.ShareType == 0 || s.ShareType == 1 {
+		return &ShareOrLink{
```

Comment:
Not asking to change, but wouldn't it there be a way to just return "interface" and then use reflection to check the type of object when doing the "create if"?

Reply:
In principle, yes, would indeed be possible! Although that seemed to be a bit too convoluted at the time compared to this

Reply:
And I think Go reflection is not so straightforward (as e.g. in Python)... so better to keep this simple. After all, this is a one-off tool, which is not part of the service.